### PR TITLE
feat: split batch for pipefusion parallelism

### PR DESCRIPTION
**Pull request by Eigensystem**
* Lossless accelerated inference by splitting hidden state in batch dimension, which is 2 because of Classifier-free guidance.

## Inference speed
* pipefusion with 8 cards, 40 steps: **13.07it/s**
* pipefusion + split batch with 8 cards, splitting hidden state into 2 batch, running on 2 pipeline, each consisting of 4 cards, 40 steps: **22.08it/s**
* accelerate 1.7x on the same number of devices

## Image quality
* pipefusion parallelism with 8 cards, full pipeline.
dog_pipefusion_hw_1024_sync_corrected_async_gn_u0_w8_mb2_warm1

* pipefusion parallelism with 8 cards, split batch.
dog_pipefusion_hw_1024_sync_corrected_async_gn_u0_w8_mb2_warm1_split_batch

### CODE REVIEW

File: scripts/pixart_example.py [hunk 1]
```diff
@@ -203,6 +211,8 @@ def main():
             num_inference_steps=args.num_inference_steps,
             output_type=args.output_type,
         )
+        if rank == 0:
+            print(210)
```
Comment (feifeibear) [inline scripts/pixart_example.py]: remove this line

File: pipefuser/utils.py [hunk 2]
```diff
@@ -115,6 +115,14 @@ def __init__(
...
         self.use_resolution_binning = use_resolution_binning
 
+        self.height = height
+        self.width = width
+
+        local_rank = local_rank = int(os.getenv("LOCAL_RANK", "0"))
+        device = torch.device(f"cuda:{local_rank}")
+        torch.cuda.set_device(device)
+        self.device = device
+
```
Comment (feifeibear) [inline pipefuser/utils.py:128]: name n_device_per_batch as pipefusion_degree?

File: pipefuser/utils.py [hunk 3]
```diff
@@ -187,16 +191,38 @@ def __init__(self, distri_config: DistriConfig):
...
-        ) % distri_config.world_size
         self.dtype = None
-        if distri_config.world_size == 2:
-            self.groups = [torch.distributed.new_group(), torch.distributed.new_group()]
+
+
+        batch_world_size = dist.get_world_size()
+        self.batch_group = dist.new_group()
+
+        if distri_config.batch_group is not None and distri_config.split_group is not None:
```
Comment (feifeibear) [inline pipefuser/utils.py]: 1. Use "split batch" by setting the dp degree and pipefusion degree
2. name the group for pipefusion as pipefusion_group. (split_group->pipefusion_group)
3. name the group for batch as dp_group. (batch_group -> dp_group)

Comment (Eigensystem) [reply pipefuser/utils.py]: It feels like the DistriConfig is getting confusing as more and more options are added, is it possible to consider separating the configs for different parallel modes into different classes afterwards to make it easier to handle?

File: pipefuser/pipelines/pip/distri_pixartalpha.py [hunk 4]
```diff
@@ -9,6 +9,8 @@
     retrieve_timesteps,
 )
 from typing import List, Optional, Union, Tuple, Callable, Dict, Final
+
+import torch.distributed
```
Comment (feifeibear) [inline pipefuser/pipelines/pip/distri_pixartalpha.py]: import torch.distributed as dist
It is not common use distributed

File: pipefuser/pipelines/pip/distri_pixartalpha.py [hunk 5]
```diff
@@ -457,7 +478,7 @@ def __call__(
                 ):
                     progress_bar.update()
 
-            if distri_config.rank == 0:
+            if torch.distributed.get_rank() == 0:
```
Comment (feifeibear) [inline pipefuser/pipelines/pip/distri_pixartalpha.py]: not necessary

File: pipefuser/utils.py [hunk 6]
```diff
@@ -166,15 +170,15 @@ def __init__(
 
     def batch_idx(self, rank: Optional[int] = None) -> int:
         if rank is None:
-            rank = self.rank
+            rank = dist.get_rank()
```
Comment (feifeibear) [inline pipefuser/utils.py:197]: not necessary

File: pipefuser/utils.py (hunk 3 above)
Comment (feifeibear) [reply pipefuser/utils.py]: OK, this PR, we can use the split batch flag. We can change the config in the future.

**Review by feifeibear — APPROVED** [review]

**Pull request merged**